README.md
# Borehole Management System Service

## Installation

```bash
sudo apt-get install \
    unzip \
    python3.7 \
    python3-pip \
    nginx \
    nano \
    npm \
    git \
    postgresql \
    postgresql-contrib \
    postgis \
    libsqlite3-mod-spatialite 
```

### Database initilization

Create the database

```bash
sudo -u postgres createdb -E UTF8 bdms
```

Create the tables:

```bash
sudo -u postgres psql -d bdms -f db/1_schema.sql
```

Add default data:

```bash
sudo -u postgres psql -d bdms -f db/2_data.sql
sudo -u postgres psql -d bdms -f db/3_cantons.sql
sudo -u postgres psql -d bdms -f db/4_municipalities.sql
```

Or remotly with port forwarding:

```bash
ssh -R 9432:localhost:5432 USER_NAME@IP_ADDRESS
```

And then:
```bash
psql -U postgres -d bdms -h localhost -p 9432 -f db/1_schema.sql
psql -U postgres -d bdms -h localhost -p 9432 -f db/2_data.sql
psql -U postgres -d bdms -h localhost -p 9432 -f db/3_cantons.sql
psql -U postgres -d bdms -h localhost -p 9432 -f db/4_municipalities.sql
```

## Run Server

### Setup virtual environment

```bash
sudo apt-get install python3-venv
python3.7 -m venv ./venv
source venv/bin/activate
pip install --upgrade pip
pip install -r requirements.txt
```

### Activate virtual environment

```bash
source venv/bin/activate
```

### Run Server

```bashm
python bms/main.py --pg-database=bdms
```

Config parameters:

```
--pg-database      PostgrSQL database name (default bms)
--pg-host          PostgrSQL database host (default localhost)
--pg-password      PostgrSQL user password (default postgres)
--pg-port          PostgrSQL database port (default 5432)
--pg-user          PostgrSQL database user (default postgres)
--pg-upgrade       Upgrade PostgrSQL schema (default false)
--port             Tornado Web port (default 8888)
```

## Run with docker

### Installation

Docker install:

```bash
sudo apt-get remove docker docker-engine docker.io containerd runc
sudo apt-get update
sudo apt-get install \
    apt-transport-https \
    ca-certificates \
    curl \
    gnupg-agent \
    software-properties-common
curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
sudo add-apt-repository \
   "deb [arch=amd64] https://download.docker.com/linux/ubuntu \
   $(lsb_release -cs) \
   stable"
sudo apt-get update
sudo apt-get install docker-ce docker-ce-cli containerd.io
```

Git install:

```bash
sudo apt-get install git
```

### Build image and run docker

Clone server repository:

```bash
git clone https://github.com/geoadmin/service-bdms.git
cd service-bdms
```

Build Docker image (x.x.x is the release number, start with 1.0.0):

```bash
 sudo docker build -t swisstopo/service-bdms:x.x.x .
```

Run Docker container (x.x.x is the release number, start with 1.0.0):

N.B. Run with Docker-compose is preferred, please see docker-compose.yml.

```bash
sudo docker run -d --name service-bdms swisstopo/service-bdms:x.x.x
```

Run Docker container with bash in interactive mode (during dev):

```bash
sudo docker run -it --name service-bdms swisstopo/service-bdms:x.x.x /bin/bash
```

CHANGES.md
# Changelog

## 1.0.2

### New features


 - **Borehole identifiers multilanguage**

    - Identifiers can be created and modified in each supported language.

 - **Text content management**

    - This feature is developed to handle multilanguage text (title and body) contents. Currently this feature is used to manage the text displayed in the login screen of web-bdms. (some default text is added in the databse)

 - **Handling database update process on upgrade**

    - The service will check the current version of the database and apply the required changes automatically if the 

## 1.0.1

### New features

 - **Feedback**: 
   
   Added new REST actions to handle user feedbacks.

 - **Terms of service**: 
   
   Added new REST actions to handle user agreements on *terms of service*.

 - **Login as viewer**: 
   
   Guest users can now access with the VIEW role.

 - **Attachments**: 
   
   You can now upload files attachments for each borehole on a AWS S3 Bucket.

 - **New configuration options**:
   
   - To configure the file repository some options have been added to configure the AWS S3 Bucket (file_repo, aws_bucket, aws_credentials)

   - feedback can be send throught and SMTP server to a configured email address.

 - **New REST API**:

   */api/v1/feedback*

   - ACTION=CREATE logged-in user can send a feedback message

   */api/v1/terms*

   - ACTION=ACCEPT logged-in user accept the current *terms of service*
   - ACTION=GET get the currently valid *terms of service*.

   */api/v1/terms/admin* (admin users only)

   - ACTION=GET Get the current draft *terms of service*.
   - ACTION=DRAFT Admins can update the current *terms of service* flagging it as *draft*.
   - ACTION=PUBLISH Promote the draft of *terms of service* to go public.

   */api/v1/borehole*

   - ACTION=LISTFILES list all files with less metadata linked to a given borehole

   */api/v1/borehole/edit*

   - ACTION=LISTFILES list all files metadata linked to a given borehole 
   
   */api/v1/borehole/edit/files*

   - ACTION=ATTACHFILE (multipart/form-data) and attach files to existing boreholes
   - ACTION=DETACHFILE unlink an uploaded file from a borehole
   - ACTION=PATCH update metadata of a file


 - **Database changes**:
   
   To handle the new attachments feature, two new tables were added to the
   schema bdms: **files** and **borehole_files**

 - **Simple Import with CSV**:
   
   As editor, importing new data using csv now handles more fields. In particular you can import easily a list of boreholes using this columns descriptors:

   - **location_east** (*numeric*): **mandatory**
   - **location_north** (*numeric*): **mandatory**
   - **public_name** (*text*): **mandatory**
   - location_z (*numeric*): *optional*
   - original_name (*text*): optional
   - project_name (*text*): optional
   - drilling_date (*text*): optional
   - total_depth (*numeric*): optional
   - top_bedrock (*numeric*): optional
   - remarks (*text*): optional

   Note: The column order can be exchanged.

### Backwards incompatible changes

 - **REST API Changes**:
   
   CSV importer API path has been changed from '/api/v1/borehole/upload' to
   '/api/v1/borehole/edit/import' and the POST-ed action param changed from
   'UPLOAD' to 'IMPORTCSV'

### Update from 1.0.0 to 1.0.0-RC1

Install boto3 in your python virtual env

```bash
source venv/bin/activate
```

Upgrade the postgresql schema by running this SQL script:

```bash
psql -U postgres -d bdms \
 -h localhost -p 9432 \
 -f assets/sql/1.0.0_to_1.0.1-RC1.sql
```


bms/__init__.py
# -*- coding: utf-8 -*-

VIEW = 0
EDIT = 1
CONTROL = 2
VALID = 3
PUBLIC = 4

from bms.v1.basehandler import BaseHandler
from bms.v1.exceptions import BmsException
from bms.v1.exceptions import BmsDatabaseException
from bms.v1.exceptions import DatabaseVersionMissmatch
from bms.v1.exceptions import AuthenticationException
from bms.v1.exceptions import AuthorizationException
from bms.v1.exceptions import WorkgroupFreezed
from bms.v1.exceptions import Locked
from bms.v1.exceptions import NotFound
from bms.v1.exceptions import MissingParameter
from bms.v1.exceptions import DuplicateException

# Borehole's ACTION Handlers
from bms.v1.borehole.producer import BoreholeProducerHandler
from bms.v1.borehole.viewer import BoreholeViewerHandler
from bms.v1.borehole.filehandler import FileHandler

# Identifiers's ACTION Handlers
from bms.v1.borehole.identifier import CreateIdentifier
from bms.v1.borehole.identifier import CreateIdentifier
from bms.v1.borehole.identifier import DeleteIdentifier
from bms.v1.borehole.identifier import PatchIdentifier 
from bms.v1.borehole.identifier import ListIdentifiers 
from bms.v1.borehole.identifier import IdentifierProducerHandler
from bms.v1.borehole.identifier import IdentifierAdminHandler
from bms.v1.borehole.identifier import IdentifierViewerHandler

# Stratigraphy's ACTION Handlers
from bms.v1.borehole.stratigraphy.viewer import StratigraphyViewerHandler
from bms.v1.borehole.stratigraphy.producer import StratigraphyProducerHandler

# ACTION Handlers exports
from bms.v1.borehole.export.exporter import ExportHandler

# Stratigraphy's layers handlers
from bms.v1.borehole.stratigraphy.layer.producer import LayerProducerHandler
from bms.v1.borehole.stratigraphy.layer.viewer import LayerViewerHandler

from bms.v1.setting.handler import SettingHandler
from bms.v1.setting.download import DownloadHandler

from bms.v1.borehole.project.handler import ProjectHandler
from bms.v1.borehole.codelist.handler import CodeListHandler
from bms.v1.geoapi.handler import GeoapiHandler
from bms.v1.geoapi.municipality.handler import MunicipalityHandler
from bms.v1.geoapi.canton.handler import CantonHandler

# Workflows handlers
from bms.v1.workflow.producer import WorkflowProducerHandler

# Content handlers
from bms.v1.content.handler import ContentHandler
from bms.v1.content.admin import ContentAdminHandler

# Terms handlers
from bms.v1.terms.handler import TermsHandler
from bms.v1.terms.admin import TermsAdminHandler

# Feedback handlers
from bms.v1.feedback.handler import FeedbackHandler

# Actions
from bms.v1.borehole import CreateBorehole
from bms.v1.borehole import ListBorehole
from bms.v1.borehole import BoreholeIds
from bms.v1.borehole import GetBorehole
from bms.v1.borehole import CheckBorehole
from bms.v1.borehole import PatchBorehole

# GeoApi actions
from bms.v1.geoapi import ListMunicipality
from bms.v1.geoapi import ListCanton
from bms.v1.geoapi import Wmts
from bms.v1.geoapi import Wms
# from bms.v1.geoapi import GetFeature

# User actions
from bms.v1.user.handler import UserHandler
from bms.v1.user.admin import AdminHandler
from bms.v1.user import ListUsers
from bms.v1.user import CheckUsername
from bms.v1.user import CreateUser

# Workgroup actions
from bms.v1.user.workgrpup.admin import WorkgroupAdminHandler
from bms.v1.user.workgrpup import ListWorkgroups
from bms.v1.user.workgrpup import CreateWorkgroup


bms/v1/exceptions.py
# -*- coding: utf-8 -*-


class BmsException(Exception):
    def __init__(self, message, code, data = None):
        super().__init__(message)
        self.code = code
        self.data = data


class BmsDatabaseException(BmsException):
    def __init__(self, *arg, **args):
        super().__init__(*arg, **args)

class DatabaseVersionMissmatch(BmsDatabaseException):
    def __init__(self, source_version, database_version):
        super().__init__(
            f"Database version (v{database_version}) missmatch. service-bdms requires (v{source_version})",
            'E-010'
        )

class ActionWrong(BmsException):
    def __init__(self):
        super().__init__("Server error", 'E-000')


class NotFound(BmsException):
    def __init__(self):
        super().__init__("Not found", 'E-050')


class ActionEmpty(BmsException):
    def __init__(self):
        super().__init__("Action empty", 'E-200')


class AuthenticationException(BmsException):
    def __init__(self):
        super().__init__("Authentication error", 'E-102')


class AuthorizationException(BmsException):
    def __init__(self):
        super().__init__("Authorization error", 'E-101')


class WorkgroupFreezed(BmsException):
    def __init__(self):
        super().__init__("Workgroup is freezed", 'E-103')


class DuplicateException(BmsException):
    def __init__(self):
        super().__init__("Record already exists", 'E-104')


class PatchAttributeException(BmsException):
    def __init__(self, attribute):
        super().__init__(f"Attribute \"{attribute}\" unknown", 'E-201')


class MissingParameter(BmsException):
    def __init__(self, parameter):
        super().__init__(f"Missing parameter {parameter}", 'E-203')


class WrongParameter(BmsException):
    def __init__(self, parameter):
        super().__init__(f"Wrong parameter {parameter}", 'E-204')


# Raised when trying to delete row from a referecnced table
class DeleteReferenced(BmsException):
    def __init__(self):
        super().__init__("Foreign key constraint violation", 'E-205')


class Locked(BmsException):
    def __init__(self, id, data):
        super().__init__(
            f"Borehole: '{id}' locked.",
            'E-900', data
        )

bms/main.py
# -*- coding: utf-8 -*-
"""This is the entry point to run the BDMS as a TornadoWeb service.
"""

__author__ = 'Institute of Earth science - SUPSI'
__version__ = '1.0.2'

from tornado import web
from tornado.options import define, options
from tornado.platform.asyncio import AsyncIOMainLoop
import asyncio
import asyncpg
from tornado.httpserver import HTTPServer
import sys
from pathlib import Path
import configparser

sys.path.append('.')

from bms.v1.listeners import EventListener

define("port", default=8888, help="Tornado Web port", type=int)
define("pg_user", default="postgres", help="PostgrSQL database user")
define("pg_password", default="postgres", help="PostgrSQL user password")
define("pg_host", default="localhost", help="PostgrSQL database host")
define("pg_port", default="5432", help="PostgrSQL database port")
define("pg_database", default="bms", help="PostgrSQL database name")
define("pg_auto_upgrade", default=False, help="Upgrade PostgrSQL schema", type=bool)

define("file_repo", default='s3', help="Select the file repository", type=str)

# Local storage for files configuration
define("local_path", default=str(Path.home()), help="Select local path", type=str)

# AWS S3 storage for files configuration
define("aws_bucket", default=None, help="Select AWS Bucket name", type=str)
define("aws_credentials", default=None, help="AWS S3 credential file location (overwrite aws_access_key_id and aws_secret_access_key)", type=str)
define("aws_access_key_id", default=None, help="AWS S3 access key id", type=str)
define("aws_secret_access_key", default=None, help="AWS S3 secret access key", type=str)

# SMTP send mail configuration
define(
    "smtp_config",
    default=None,
    help="SMTP configuration file location",
    type=str
)
define(
    "smtp_recipients",
    default=None,
    help="SMTP comma separated recipients email addresses",
    type=str
)
define(
    "smtp_username",
    default=None,
    help="SMTP username",
    type=str
)
define(
    "smtp_password",
    default=None,
    help="SMTP password",
    type=str
)
define(
    "smtp_server",
    default=None,
    help="SMTP server address",
    type=str
)
define(
    "smtp_port",
    default=587,
    help="SMTP server port",
    type=int
)
define(
    "smtp_tls",
    default=False,
    help="SMTP server supports direct connection via TLS/SSL",
    type=bool
)
define(
    "smtp_starttls",
    default=True,
    help="SMTP servers support the STARTTLS extension",
    type=bool
)


# Ordered list of upgradable versions
versions = [
    "1.0.0",
    "1.0.1",
]

# SQL upgrades directory
udir = "./bms/assets/sql/"

# SQL to execute for upgrades
sql_files = {
    "1.0.0": f"{udir}1.0.0_to_1.0.1.sql",
    "1.0.1": f"{udir}1.0.1_to_1.0.2.sql",
}

listeners = []

async def get_conn():
    try:
        return await asyncpg.create_pool(
            user=options.pg_user,
            password=options.pg_password,
            database=options.pg_database,
            host=options.pg_host,
            port=options.pg_port
        )
    except Exception as x:
        red("Connection to db failed.")
        raise x

def red(message):
    print(f"\033[91m{message}\033[0m")

def green(message):
    print(f"\033[92m{message}\033[0m")

def blue(message):
    print(f"\033[94m{message}\033[0m")

async def upgrade_database(pool):
    async with pool.acquire() as conn:
        try:
            await conn.execute("BEGIN;")
            current_db_version = await conn.fetchval("""
                SELECT
                    value_cfg
                FROM
                    bdms.config
                WHERE
                    name_cfg = 'VERSION';
            """)

            print("\nUpgrading database:")
            for idx in range(
                versions.index(current_db_version),
                len(versions)
            ):
                version = versions[idx]
                with open(sql_files[version]) as sql_file:
                    print(f" - Executing: {sql_files[version]}")
                    await conn.execute(sql_file.read())

            # Update current datetime of this upgrade
            await conn.execute("""
                UPDATE
                    bdms.config
                SET
                    value_cfg = to_char(
                        now(), 'YYYY-MM-DD"T"HH24:MI:SSOF'
                    )
                WHERE
                    name_cfg = 'PG_UPGRADE';
            """)
            print(f" - Updated upgrade date")

            # Update previous version
            await conn.execute("""
                UPDATE
                    bdms.config
                SET
                    value_cfg = $1
                WHERE
                    name_cfg = 'PREVIOUS';
            """, current_db_version)

            # Update current version
            await conn.execute("""
                UPDATE
                    bdms.config
                SET
                    value_cfg = $1
                WHERE
                    name_cfg = 'VERSION';
            """, __version__)

            await conn.execute("COMMIT;")

            print(f"\n 😃 \033[92mDatabase upgraded to latest version (v{__version__})\033[0m\n")

        except Exception as ex:
            red("\n 😞 Sorry, an error occured during the upgrade process.\n")
            await conn.execute("ROLLBACK;")
            raise ex

async def system_check(pool):
    async with pool.acquire() as conn:
        # Checking database version
        current_db_version = await conn.fetchval("""
            SELECT
                value_cfg
            FROM
                bdms.config
            WHERE
                name_cfg = 'VERSION';
        """)

    if current_db_version != __version__:
        # Raise exception if database version missmatch
        #   with service-bdms requirements
        from bms import DatabaseVersionMissmatch
        raise DatabaseVersionMissmatch(
            __version__,
            current_db_version
        )

async def close(application):
    # Remove all listeners
    blue("Removing listeners..")
    await application.listener.stop()
    green(" > done.")

    blue("Closing connection pool..")
    # Closing connections pool
    await application.pool.close()
    green(" > done.")

    print("")

if __name__ == "__main__":

    options.parse_command_line()

    from bms import (
        # Exceptions
        BmsDatabaseException,
        DatabaseVersionMissmatch,

        # user handlers
        SettingHandler,
        DownloadHandler,
        UserHandler,
        AdminHandler,
        WorkgroupAdminHandler,

        # Borehole handlers
        BoreholeViewerHandler,
        BoreholeProducerHandler,
        # BoreholeExporterHandler,
        ExportHandler,
        FileHandler,

        # Identifier handlers
        IdentifierAdminHandler,
        IdentifierProducerHandler,
        IdentifierViewerHandler,

        # Stratigraphy handlers
        StratigraphyViewerHandler,
        StratigraphyProducerHandler,

        # Layer handlers
        LayerViewerHandler,
        LayerProducerHandler,

        # Workflow handlers
        WorkflowProducerHandler,

        # Content handlers
        ContentHandler,
        ContentAdminHandler,

        # Terms handlers
        TermsHandler,
        TermsAdminHandler,

        # Feedback handler
        FeedbackHandler,

        # Other handlers
        GeoapiHandler,
        ProjectHandler,
        CodeListHandler,
        MunicipalityHandler,
        CantonHandler,
        Wmts,
        Wms,
        # GetFeature
    )

    AsyncIOMainLoop().install()
    ioloop = asyncio.get_event_loop()
    ioloop.set_debug(enabled=True)

    settings = dict(
        debug=True
    )

    application = web.Application([

        # Borehole handlers
        (r'/api/v1/setting', SettingHandler),
        (r'/api/v1/setting/download', DownloadHandler),
        (r'/api/v1/user', UserHandler),
        (r'/api/v1/user/edit', AdminHandler),

        (r'/api/v1/user/workgroup/edit', WorkgroupAdminHandler),

        # Borehole handlers
        (r'/api/v1/borehole', BoreholeViewerHandler),
        (r'/api/v1/borehole/edit', BoreholeProducerHandler),
        (r'/api/v1/borehole/download', ExportHandler),
        (r'/api/v1/borehole/edit/import', BoreholeProducerHandler),
        (r'/api/v1/borehole/edit/files', FileHandler),

        # Stratigraphy handlers
        (r'/api/v1/borehole/identifier', IdentifierViewerHandler),
        (r'/api/v1/borehole/identifier/edit', IdentifierProducerHandler),
        (r'/api/v1/borehole/identifier/admin', IdentifierAdminHandler),

        # Workflow handlers
        (r'/api/v1/workflow/edit', WorkflowProducerHandler),

        # Content handlers
        (r'/api/v1/content/admin', ContentAdminHandler),
        (r'/api/v1/content', ContentHandler),
        (r'/api/v1/content/(.*)', ContentHandler),

        # Terms handlers
        (r'/api/v1/terms', TermsHandler),
        (r'/api/v1/terms/admin', TermsAdminHandler),

        # FEEDBACK handlers
        (r'/api/v1/feedback', FeedbackHandler),

        # Stratigraphy handlers
        (r'/api/v1/borehole/stratigraphy', StratigraphyViewerHandler),
        (r'/api/v1/borehole/stratigraphy/edit', StratigraphyProducerHandler),

        # Layer handlers
        (r'/api/v1/borehole/stratigraphy/layer', LayerViewerHandler),
        (r'/api/v1/borehole/stratigraphy/layer/edit', LayerProducerHandler),

        # Other handlers
        # (r'/api/v1/borehole/project', ProjectHandler),
        (r'/api/v1/borehole/codes', CodeListHandler),
        (r'/api/v1/geoapi/municipality', MunicipalityHandler),
        (r'/api/v1/geoapi/canton', CantonHandler),
        (r'/api/v1/geoapi/location', GeoapiHandler),
        (r"/api/v1/geoapi/wmts", Wmts),
        (r"/api/v1/geoapi/wms/swisstopo", Wms),
        # (r"/api/v1/geoapi/getfeature", GetFeature)

    ], **settings)

    # Init database postgresql connection pool
    application.pool = ioloop.run_until_complete(get_conn())

    # Create events listeners
    application.listener = EventListener(application)
    ioloop.run_until_complete(application.listener.start())

    # Init config file parser
    config = configparser.ConfigParser()

    # Configuring S3 credentials
    if options.file_repo == 's3': # and options.aws_credentials is not None:

        # if options.aws_credentials is None:
        #     raise Exception("AWS Credential file missing")

        # File with credentials
        if options.aws_credentials is not None:
        
            config.read(options.aws_credentials)

            if (
                'default' not in config or
                'aws_access_key_id' not in config['default'] or
                'aws_secret_access_key' not in config['default']
            ):
                raise Exception("AWS Credential wrong")

            options.aws_access_key_id = config['default']['aws_access_key_id']
            options.aws_secret_access_key = config['default']['aws_secret_access_key']

        # Validate mandatory Bucket name
        if options.aws_bucket is None:
            raise Exception("AWS Bucket name (--aws_bucket) is missing")

    # Configuring SMTP credentials
    if options.smtp_config is not None:
        
        config.read(options.smtp_config)

        if (
            'SMTP' not in config or
            'smtp_recipients' not in config['SMTP'] or
            'smtp_username' not in config['SMTP'] or
            'smtp_password' not in config['SMTP'] or
            'smtp_server' not in config['SMTP']
        ):
            raise Exception("SMTP config file wrong")

        options.smtp_recipients = config['SMTP']['smtp_recipients']
        options.smtp_username = config['SMTP']['smtp_username']
        options.smtp_password = config['SMTP']['smtp_password']
        options.smtp_server = config['SMTP']['smtp_server']

        if 'smtp_port' in config['SMTP']:
            options.smtp_port = int(config['SMTP']['smtp_port'])

        if 'smtp_tls' in config['SMTP']:
            options.smtp_tls = (
                True
                if config['SMTP']['smtp_tls'] == '1'
                else False
            )

        if 'smtp_starttls' in config['SMTP']:
            options.smtp_starttls = (
                True
                if config['SMTP']['smtp_starttls'] == '1'
                else False
            )

    try:
        # Check system before startup
        try:
            ioloop.run_until_complete(
                system_check(application.pool)
            )

        except DatabaseVersionMissmatch as dvm:

            # Upgrade the database automatically
            if options.pg_auto_upgrade is True:

                ioloop.run_until_complete(
                    upgrade_database(application.pool)
                )

            else:
                raise dvm

        http_server = HTTPServer(application)
        http_server.listen(options.port)
        ioloop.run_forever()

    except DatabaseVersionMissmatch as dvm:
        print(f"""
\033[91m{dvm}\033[0m

Run this main.py script adding the --pg-auto-upgrade parameter
to upgrade your database automatically.

Be sure to have a backup of your data before proceding.
""")

    except BmsDatabaseException as du:
        print(f"\n 😃 \033[92m{du}\033[0m\n")

    except Exception as ex:
        print(f"Exception:\n{ex}")

    except KeyboardInterrupt:
        red("\nKeyboard interruption, probably: CTR+C\n")

    finally:
        ioloop.run_until_complete(close(application))
